query-algebrizer/src/cc.rs

```diff
@@ -0,0 +1,795 @@
+/// !   * Subqueries. Perhaps less efficient? Certainly clearer.
+/// !   * Projection expressions, if only used for output.
+/// !   * Inline expressions?
+/// !---------------------------------------------------------------------------------------
+/// !
+/// ! `from` is a list of [source alias] pairs, suitable for passing to honeysql.
+/// ! `bindings` is a map from var to qualified columns.
+/// ! `known-types` is a map from var to type keyword (e.g., :db.type/ref)
+/// ! `extracted-types` is a mapping, similar to `bindings`, but used to pull
+/// ! type tags out of the store at runtime.
+/// ! `wheres` is a list of fragments that can be joined by `:and`.
+
+extern crate mentat_core;
+extern crate mentat_query;
+
+use ::std::collections::BTreeMap;
+use ::std::collections::btree_map::Entry;
+
+use self::mentat_core::{
+    Attribute,
+    Entid,
+    Schema,
+    TypedValue,
+    ValueType,
+};
+
+use self::mentat_query::{
+    IdentOrEntid,
+    NamespacedKeyword,
+    NonIntegerConstant,
+    Pattern,
+    PatternNonValuePlace,
+    PatternValuePlace,
+    PlainSymbol,
+    SrcVar,
+    Variable,
+};
+
+/// This enum models the fixed set of default tables we have -- two
+/// tables and two views.
+#[derive(PartialEq, Eq, Clone, Copy, Debug)]
+pub enum DatomsTable {
+    Datoms,             // The non-fulltext datoms table.
+    FulltextValues,     // The virtual table mapping IDs to strings.
+    FulltextDatoms,     // The fulltext-datoms view.
+    AllDatoms,          // Fulltext and non-fulltext datoms.
+}
+
+impl DatomsTable {
+    fn name(&self) -> &'static str {
+        match *self {
+            DatomsTable::Datoms => "datoms",
+            DatomsTable::FulltextValues => "fulltext_values",
+            DatomsTable::FulltextDatoms => "fulltext_datoms",
+            DatomsTable::AllDatoms => "all_datoms",
+        }
+    }
+}
+
+/// One of the named columns of our tables.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub enum DatomsColumn {
+    Entity,
+    Attribute,
+    Value,
+    Tx,
+    ValueTypeTag,
+}
+
+/// A specific instance of a table within a query. E.g., "datoms123".
+pub type TableAlias = String;
+
+/// The association between a table and its alias. E.g., AllDatoms, "all_datoms123".
+#[derive(PartialEq, Eq, Debug)]
+pub struct SourceAlias(DatomsTable, TableAlias);
+
+/// A particular column of a particular aliased table. E.g., "datoms123", Attribute.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct QualifiedAlias(TableAlias, DatomsColumn);
+
+impl QualifiedAlias {
+    fn for_type_tag(&self) -> QualifiedAlias {
+        QualifiedAlias(self.0.clone(), DatomsColumn::ValueTypeTag)
+    }
+
+    #[inline]
+    fn is_value(&self) -> bool {
+        self.1 == DatomsColumn::Value
+    }
+}
+
+/// A thing that's capable of aliasing a table name for us.
+/// This exists so that we can obtain predictable names in tests.
+pub type TableAliaser = Box<FnMut(DatomsTable) -> TableAlias>;
```

Comment:
This thing really only exists to get predictable names, and I might just replace it with a mutable counter in the CC itself.